velox/functions/sparksql/tests/FromJsonTest.cpp

```diff
@@ -0,0 +1,249 @@
+       R"({"a": "+INF"})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, basicDouble) {
+  auto expected = makeNullableFlatVector<double>(
+      {1.0,
+       2.0,
+       std::nullopt,
+       kNaNDouble,
+       -kInfDouble,
+       -kInfDouble,
+       kInfDouble,
+       kInfDouble,
+       kInfDouble});
+  auto input = makeFlatVector<std::string>(
+      {R"({"a": 1})",
+       R"({"a": 2.0})",
+       R"({"a": "3"})",
+       R"({"a": "NaN"})",
+       R"({"a": "-Infinity"})",
+       R"({"a": "-INF"})",
+       R"({"a": "+Infinity"})",
+       R"({"a": "Infinity"})",
+       R"({"a": "+INF"})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, basicString) {
+  auto expected = makeNullableFlatVector<StringView>(
+      {"1", "2.0", "true", "{\"b\": \"test\"}", "[1, 2]"});
+  auto input = makeFlatVector<std::string>(
+      {R"({"a": 1})",
+       R"({"a": 2.0})",
+       R"({"a": "true"})",
+       R"({"a": {"b": "test"}})",
+       R"({"a": [1, 2]})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, nestedComplexType) {
+  auto rowVector = makeRowVector({"a"}, {makeFlatVector<int64_t>({1, 2, 2})});
+  std::vector<vector_size_t> offsets;
+  offsets.push_back(0);
+  offsets.push_back(1);
+  offsets.push_back(2);
+  auto arrayVector = makeArrayVector(offsets, rowVector);
+  auto input = makeFlatVector<std::string>(
+      {R"({"a": 1})", R"([{"a": 2}])", R"([{"a": 2}])"});
+  testFromJson(input, arrayVector);
+}
+
+TEST_F(FromJsonTest, nullOnFailure) {
+  auto expected = makeNullableFlatVector<int64_t>({1, std::nullopt, 3});
+  auto input =
+      makeFlatVector<std::string>({R"({"a": 1})", R"({"a" 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, structEmptyArray) {
+  auto expected = makeNullableFlatVector<int64_t>({std::nullopt, 2, 3});
+  auto input =
+      makeFlatVector<std::string>({R"([])", R"({"a": 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, structEmptyStruct) {
+  auto expected = makeNullableFlatVector<int64_t>({std::nullopt, 2, 3});
+  auto input =
+      makeFlatVector<std::string>({R"({ })", R"({"a": 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, structWrongSchema) {
+  auto expected = makeNullableFlatVector<int64_t>({std::nullopt, 2, 3});
+  auto input = makeFlatVector<std::string>(
+      {R"({"b": 2})", R"({"a": 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, structWrongData) {
+  auto expected = makeNullableFlatVector<int64_t>({std::nullopt, 2, 3});
+  auto input = makeFlatVector<std::string>(
+      {R"({"a": 2.1})", R"({"a": 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, invalidType) {
```

Comment:
Would you also add tests for invalid  JSON input?

Reply:
Updated, thanks.